third_party/libtock-drivers/src/nfc.rs

```diff
@@ -76,61 +76,53 @@ impl NfcTag {
     /// 1. Share with the driver a buffer.
     /// 2. Subscribe to having a successful receive callback.
     /// 3. Issue the request for reception.
-    pub fn receive(buf: &mut [u8; 256]) -> TransOrRecvStatus {
-        let result = syscalls::allow(DRIVER_NUMBER, allow_nr::RECEIVE, buf);
-        if result.is_err() {
-            return TransOrRecvStatus::InvalidBuffer;
-        }
-
+    pub fn receive(buf: &mut [u8; 256]) -> Result<RecvOp, TockError> {
+        let result = syscalls::allow(DRIVER_NUMBER, allow_nr::RECEIVE, buf)?;
+        // set callback with 2 arguments, to receive ReturnCode and RX Amount
         let done = Cell::new(false);
-        let mut alarm = || done.set(true);
-        let subscription = syscalls::subscribe::<callback::Identity0Consumer, _>(
+        let result_code = Cell::new(None);
+        let recv_amount = Cell::new(None);
+        let mut callback = |result, amount| {
+            result_code.set(Some(result));
+            recv_amount.set(Some(amount));
+            done.set(true)
+        };
+        let subscription = syscalls::subscribe::<callback::Identity2Consumer, _>(
             DRIVER_NUMBER,
             subscribe_nr::RECEIVE,
-            &mut alarm,
-        );
-        if subscription.is_err() {
-            return TransOrRecvStatus::Error;
-        }
-
-        let result_code =
-            unsafe { syscalls::raw::command(DRIVER_NUMBER, command_nr::RECEIVE, 0, 0) };
-        match result_code {
-            SUCCESS => (),
-            ENOMEM => return TransOrRecvStatus::OOM,
-            _ => return TransOrRecvStatus::Error,
-        }
-
+            &mut callback,
+        )?;
+        syscalls::command(DRIVER_NUMBER, command_nr::RECEIVE, 0, 0)?;
         util::yieldk_for(|| done.get());
-        TransOrRecvStatus::Success
+        mem::drop(subscription);
+        mem::drop(result);
+        Ok(RecvOp {
+            result_code: result_code.get().unwrap(),
+            recv_amount: recv_amount.get().unwrap(),
+        })
     }
 
     /// 1. Share with the driver a buffer containing the app's reply.
     /// 2. Subscribe to having a successful transmission callback.
     /// 3. Issue the request for transmitting.
-    pub fn transmit(buf: &mut [u8], amount: usize) -> TransOrRecvStatus {
-        let result = syscalls::allow(DRIVER_NUMBER, allow_nr::TRANSMIT, buf);
-        if result.is_err() {
-            return TransOrRecvStatus::InvalidBuffer;
-        }
-
+    pub fn transmit(buf: &mut [u8], amount: usize) -> Result<usize, TockError> {
+        let result = syscalls::allow(DRIVER_NUMBER, allow_nr::TRANSMIT, buf)?;
+        // set callback with 1 argument, to receive ReturnCode
         let done = Cell::new(false);
-        let mut alarm = || done.set(true);
-        let subscription = syscalls::subscribe::<callback::Identity0Consumer, _>(
+        let result_code = Cell::new(None);
```

Comment:
Your `Cell` contains an option, so you don't need the extra `bool` as an indicator? So if you don't want to keep the `done` pattern, you could replace the `dont.get()` check below? Same for `receive`.